src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,100 +152,125 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
-                bool targetNeedsRestore;
-                ProjectLoadTelemetryReporter.TelemetryInfo targetTelemetryInfo;
+                var loadedProjectInfos = await loadedFile.GetProjectFileInfosAsync(cancellationToken);
 
-                if (existingProject != null)
+                // The out-of-proc build host supports more languages than we may actually have Workspace binaries for, so ensure we can actually process that
+                // language in-process.
+                var projectLanguage = loadedProjectInfos.FirstOrDefault()?.Language;
+                if (projectLanguage != null && ProjectFactory.Workspace.Services.GetLanguageService<ICommandLineParserService>(projectLanguage) == null)
                 {
-                    (targetTelemetryInfo, targetNeedsRestore) = await existingProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+                    return false;
                 }
-                else
+
+                Dictionary<ProjectFileInfo, ProjectLoadTelemetryReporter.TelemetryInfo> telemetryInfos = [];
+                var needsRestore = false;
+
+                using var _ = await loadedProjectSet.Semaphore.DisposableWaitAsync(cancellationToken);
+                // last chance for someone to cancel out from under us.
+                if (loadedProjectSet.CancellationTokenSource.IsCancellationRequested)
                 {
-                    var projectSystemName = $"{projectPath} (${loadedProjectInfo.TargetFramework})";
-                    var projectCreationInfo = new ProjectSystemProjectCreationInfo
-                    {
-                        AssemblyName = projectSystemName,
-                        FilePath = projectPath,
-                        CompilationOutputAssemblyFilePath = loadedProjectInfo.IntermediateOutputFilePath
-                    };
+                    return false;
+                }
 
-                    var projectSystemProject = await ProjectFactory.CreateAndAddToWorkspaceAsync(
-                        projectSystemName,
-                        loadedProjectInfo.Language,
-                        projectCreationInfo,
-                        _projectSystemHostInfo);
+                var existingProjects = loadedProjectSet.LoadedProjects;
 
-                    var loadedProject = new LoadedProject(projectSystemProject, ProjectFactory.Workspace.Services.SolutionServices, _fileChangeWatcher, _targetFrameworkManager);
-                    loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(projectToLoad with { ReportTelemetry = false });
-                    existingProjects.Add(loadedProject);
+                // We want to remove projects for targets that don't exist anymore; if we update projects we'll remove them from  
+                // this list -- what's left we can then remove.
+                HashSet<LoadedProject> projectsToRemove = [.. existingProjects];
```

Comment:
```suggestion
                var projectsToRemove = new HashSet<LoadedProject>(existingProjects);
```

Reply:
Why?

Reply:
in this case, since we're going to say the type *anyways*, might as well just have one construct.  instead of having to state hte typ e*and* then say: make it by spreading.  Basically, i view the 'constructor' as conveying both.  This is a nit.  Can be ignored fully.